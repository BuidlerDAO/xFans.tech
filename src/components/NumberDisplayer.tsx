import React from 'react';
import BigNumber from 'bignumber.js';

/**
 * 展示 bignumber 组件
 *
 * eg. 0.0{4}5252
 */
export function NumberDisplayer({ text = '0' }: { text?: string }) {
  const number = new BigNumber(text).dividedBy(new BigNumber(Math.pow(10, 18))).toFixed();
  const [valueBeforeDot, valueAfterDot] = number.split('.');
  let value;
  if (valueAfterDot == null) {
    value = '00';
  } else if (BigNumber(valueBeforeDot).gte(1)) {
    value = BigNumber(`0.${valueAfterDot}`).toFixed(2).split('.')[1];
  } else if (
    BigNumber(valueBeforeDot).isLessThan(1) &&
    BigNumber(`0.${valueAfterDot}`).gte(0.0001)
  ) {
    // 展示完整, 四舍五入
    value = BigNumber(`0.${valueAfterDot}`).toFixed(4).split('.')[1];
  } else {
    const lastZeroIndex = valueAfterDot.lastIndexOf('0');
    const zeroCount = valueAfterDot.substring(0, lastZeroIndex + 1).length;
    const noneZeroValue = valueAfterDot.substring(lastZeroIndex + 1);
    const [, noneZeroRealValue] = (+`0.${noneZeroValue}`).toFixed(4).split('.');
    value = `0{${zeroCount}}${noneZeroRealValue}`;
  }
  return (
    <span>
      {valueBeforeDot}.{value}
    </span>
  );
}
