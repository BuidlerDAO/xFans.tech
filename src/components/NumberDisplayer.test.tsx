import { render } from '@testing-library/react';

import { NumberDisplayer } from './NumberDisplayer';

it('balance should display well', () => {
  expect(render(<NumberDisplayer text="0.000000180813661601" />).container).toMatchSnapshot();
});
it('bignumber should be transform automatically', () => {
  expect(render(<NumberDisplayer text="3368437500000000" />).container).toMatchSnapshot();
});
it('integer should be append two zero to tail', () => {
  expect(render(<NumberDisplayer text="3" />).container).toMatchSnapshot();
});
it('if interger part great than 1, decimal part should have 2 decimals', () => {
  expect(render(<NumberDisplayer text="3.145" />).container).toMatchSnapshot();
});
it('if interger part less than 1, decimal part should have 4 decimals', () => {
  expect(render(<NumberDisplayer text="0.145" />).container).toMatchSnapshot();
});
it('if interger part great than 1, decimal part should round into 4 decimals', () => {
  expect(render(<NumberDisplayer text="0.14235" />).container).toMatchSnapshot();
});

it('should throw error if text is not string', () => {
  expect(() =>
    render(<NumberDisplayer text={3.14 as unknown as string} />)
  ).toThrowErrorMatchingSnapshot();
});
