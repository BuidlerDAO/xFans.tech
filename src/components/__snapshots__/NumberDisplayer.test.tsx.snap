// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`balance should display well 1`] = `
<div>
  <span>
    0.0{6}1808
  </span>
</div>
`;

exports[`bignumber should be transform automatically 1`] = `
<div>
  <span>
    3368437500000000.00
  </span>
</div>
`;

exports[`if interger part great than 1, decimal part should have 2 decimals 1`] = `
<div>
  <span>
    3.15
  </span>
</div>
`;

exports[`if interger part great than 1, decimal part should round into 4 decimals 1`] = `
<div>
  <span>
    0.1424
  </span>
</div>
`;

exports[`if interger part less than 1, decimal part should have 4 decimals 1`] = `
<div>
  <span>
    0.1450
  </span>
</div>
`;

exports[`integer should be append two zero to tail 1`] = `
<div>
  <span>
    3.00
  </span>
</div>
`;

exports[`should throw error if text is not string 1`] = `"text should be string"`;
